assemble.go

```diff
@@ -3,11 +3,91 @@ package desync
 import (
 	"context"
 	"fmt"
+	"golang.org/x/sync/errgroup"
 	"os"
+	"sync"
+)
 
-	"golang.org/x/sync/errgroup"
+// InvalidSeedAction represent the action that we will take if a seed
+// happens to be invalid. There are currently two options: either fail with
+// an error or skip the invalid seed and try to continue.
+type InvalidSeedAction int
+
+const (
+	InvalidSeedActionBailOut InvalidSeedAction = iota
+	InvalidSeedActionSkip
 )
 
+type AssembleOptions struct {
+	N                 int
+	InvalidSeedAction InvalidSeedAction
+}
+
+// validatePlan validates a proposed plan by checking if all the chosen chunks
+// are correctly provided from the seeds. In case a seed has invalid chunks, the
+// entire seed is marked as invalid and an error is returned.
+func validatePlan(ctx context.Context, plan []SeedSegmentCandidate) (err error) {
```

Comment:
I think this is where you're losing that little bit of performance, and there's significant room for simplification in this function.
1. In seedsequencer.go, you could introduce `type Plan []SeedSegmentCandidate`
2. Then move this function to seedsequencer.go and make it a method of the Plan like so, `func (p Plan) Validate(ctx context.Context, n int)` - you'll need the `n` here for the level of concurrency. You can pass in `N` from the options
3. On the SeedSegment interface, expose the Validate method (it already exists private on the fileseed)
4. Once you have a Validate() method on every segment, you can just call that and you don't have to work with a file descriptor here at all.
5. Open more than one goroutine (loop over `n`) to actually validate the plan concurrently. The `g.Go()` should be in a loop.

Once you have that, I bet the perf difference will disappear completely, and this function will be quite simple and short. If my explanation above doesn't make sense, let me know and I can make a PR for your PR.

Reply:
> The g.Go() should be in a loop.

Oops, I forgot to create the `g.Go()` in a loop, so mine was not really doing the validation concurrently.

I implemented what you suggested and now it should be better.

However you mentioned that "you don't have to work with a file descriptor here at all", but I'm not sure if with "here" you meant in `assemble.go` or in the `func (p Plan) Validate()`.
Because to validate a particular `fileSeed` we need a filehandle, and if I open them beforehand and share them to the different Goroutines, we can drastically reduce the amount of `Open()` and `Close()` that we need.

Please let me know if you had something else in mind.

Reply:
Nope, you got it there, re-using the filehandle for concurrent validation. Looking good.